### PR TITLE
WEB-1629.3 Implement the RequestHandler base class

### PR DESCRIPTION
**Pull request by somewhatabstract**
This implements the base class and interface for request handlers as per ADR 286. See that and subsequent PRs for more.

- [x] Write tests
- [x] Update docs

### CODE REVIEW

**Comment by jeresig** [comment]
Deploy preview for *wonder-blocks* ready!

Built with commit c2b77531826ee445f2c46ad49f5ffc69ae94aabf

https://deploy-preview-619--wonder-blocks.netlify.com

File: packages/wonder-blocks-data/util/request-handler.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    invalidateCache(options: TOptions): boolean {
+        /**
+         * By default, the cache is always valid.
+         */
+        return false;
+    }
+
+    getKey(options: TOptions): string {
+        try {
+            return String(JSON.stringify(options));
```
Comment (jeresig) [inline packages/wonder-blocks-data/util/request-handler.js]: Why are you wrapping this with a `String(...)` - `JSON.stringify` already returns a string. If this is a Flow thing then just casting the value should be fine.

File: packages/wonder-blocks-data/util/request-handler.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+base class to implement your own request handler.
+
+The constructor requires a `type` to identify your handler. This should be unique
+among the handlers that are used across your application, otherwise, requests
+may be fulfilled by the wrong handler.
+
+The `fulfillRequest` method of this class is not implemented and will throw if
+called.
+
+A default implemnetation of `getKey` is provided that serializes the options of
```
Comment (jeresig) [inline packages/wonder-blocks-data/util/request-handler.md]: *implementation

File: packages/wonder-blocks-data/util/request-handler.js (hunk 1 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/util/request-handler.js]: `JSON.stringify` can return `undefined` which is not a valid string. We want an actual string here, always.

Comment (jeresig) [reply packages/wonder-blocks-data/util/request-handler.js]: It seems like this only happens if a string isn't passed in, like: `JSON.stringify(undefined)` or `JSON.stringify(function(){})`. I think in any case where `undefined` is returned we definitely want to treat that like an error and not just swallow it, converting it to a string.

File: packages/wonder-blocks-data/util/types.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+     * If this returns true, the framework will fulfill a new request by
+     * calling `fulfillRequest`.
+     */
+    invalidateCache(options: TOptions): boolean;
+
+    /**
+     * Get the key to use for a given request. This should be idempotent for a
+     * given options set if you want caching to work across requests.
+     */
+    getKey(options: TOptions): string;
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/util/types.js]: Nice when you can copy/paste straight from the ADR. :)

File: packages/wonder-blocks-data/util/request-handler.js (hunk 1 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-data/util/request-handler.js]: I changed the flow types for `JSON.stringify` in https://github.com/Khan/webapp/blob/master/flow-typed/json.js to get around this.  We could do the same here.

File: packages/wonder-blocks-data/util/request-handler.test.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    afterEach(() => {
+        /**
+         * This is needed or the JSON.stringify mocks need to be
+         * mockImplementationOnce. This is because if the snapshots need
+         * to update, they write the inline snapshot and that appears to invoke
+         * prettier which in turn, calls JSON.stringify. And if that mock
+         * throws, then boom. No snapshot update and a big old confusing test
+         * failure.
+         */
+        jest.restoreAllMocks();
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/util/request-handler.test.js]: That's pretty scary.  Thanks for writing explanation of what goes on without this call to `restoreAllMocks()`

File: packages/wonder-blocks-data/util/request-handler.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    get type() {
+        return this._type;
+    }
+
+    invalidateCache(options: TOptions): boolean {
+        /**
+         * By default, the cache is always valid.
+         */
+        return false;
+    }
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/util/request-handler.js]: Is there anytime a sub-claass of RequestHandler would also want to return `false`?

File: packages/wonder-blocks-data/util/request-handler.js (hunk 1 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/util/request-handler.js]: I don't want to block `undefined`. It's a totally valid thing for options to be if, say, a handler doesn't even have options.

I could handle it differently though.

File: packages/wonder-blocks-data/util/types.js (hunk 3 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/util/types.js]: Well, technically, some of this was copied from the ADR, then changed, then copied back. :grin:

File: packages/wonder-blocks-data/util/request-handler.js (hunk 5 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/util/request-handler.js]: Not sure I understand the question. Can you rephrase it?

Comment (somewhatabstract) [reply packages/wonder-blocks-data/util/request-handler.js]: I've updated the `request-handler.md` to cover this method and its uses.

File: packages/wonder-blocks-data/util/request-handler.js (hunk 1 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/util/request-handler.js]: I've replaced this use of `String` with a check for undefined and a cast.

File: packages/wonder-blocks-data/util/request-handler.md (hunk 2 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/util/request-handler.md]: Fixed
